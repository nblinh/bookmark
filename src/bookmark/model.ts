import { dataToBookmark } from "../utils/bookmarkUtils";

export class Bookmark {
    view: string;
    url: string;
    title: string;
    author: string;
    addDate: Date;
    publicDate: Date;
    duration?: number; //in seconde
    width?: number;
    height?: number;

    constructor(view: string, url: string, title: string, author: string, publicDate: string) {
        this.view = view;
        this.url = url;
        this.title = title;
        this.author = author;
        this.addDate = new Date();
        this.publicDate = new Date(publicDate);
    }
}

export const getBookmarkInfo = async (url: string): Promise<Bookmark> => {
    if (!url.includes('vimeo.com') && !url.includes('flickr.com')) {
        throw new Error();
    }

    return fetch('https://noembed.com/embed?url=' + url)
        .then(response => response.json())
        .then(data => dataToBookmark(data))
}