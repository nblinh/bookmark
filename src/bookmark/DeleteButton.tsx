import { FC } from "react";
import { Bookmark } from "./model";

interface DeleteButtonProps {
    bookmarks: Bookmark[];
    setBookmarks: Function;
    bookmark: Bookmark;
}

const DeleteButton: FC<DeleteButtonProps> = ({ bookmarks, setBookmarks, bookmark }) => {
    const deleteBookmark = () => {
        setBookmarks([...bookmarks.filter(item => item.url !== bookmark.url)])
    }
    return (<td style={{ border: '1px solid black' }}><input type="button" value="Supprimer" onClick={deleteBookmark} data-testid='deleteButton' /></td>);
}

export default DeleteButton;