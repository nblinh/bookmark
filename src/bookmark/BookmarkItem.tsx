import { FC } from "react";
import { getTimefromDate, secondsToHms } from "../utils/dateUtils";
import DeleteButton from "./DeleteButton";
import { Bookmark } from "./model";

interface BookmarkItemProps {
    bookmarks: Bookmark[]
    setBookmarks: Function;
    bookmark: Bookmark;
}

const BookmarkItem: FC<BookmarkItemProps> = ({ bookmarks, setBookmarks, bookmark }) => {
    return (
        <tr>
            {bookmark.duration && (
                <>
                    <td style={{ border: '1px solid black' }}><iframe title={bookmark.title} src={'https://player.vimeo.com' + bookmark.view.replace('s', '')} width='250' height='150' allow='autoplay; fullscreen; picture-in-picture'></iframe></td>
                    <td style={{ paddingBottom: '20px', border: '1px solid black' }}>
                        <table>
                            <tr><td align='left'><li><b>URL:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.url}</td></tr>
                            <tr><td align='left'><li><b>Titre de la vid√©o:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.title}</td></tr>
                            <tr><td align='left'><li><b>Auteur:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.author}</td></tr>
                            <tr><td align='left'><li><b>Date d'ajout dans l'application:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>
                                il y a {getTimefromDate(bookmark.addDate)}</td></tr>
                            <tr><td align='left'><li><b>Date de publication sur Vimeo:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>
                                <span>le {new Intl.DateTimeFormat("fr-FR", { year: "numeric", month: "long", day: "2-digit" }).format(bookmark.publicDate)}</span></td>
                            </tr>
                            <tr><td align='left'><li><b>Duration:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{secondsToHms(bookmark.duration)}</td></tr>
                        </table>
                    </td>
                    <DeleteButton bookmarks={bookmarks} setBookmarks={setBookmarks} bookmark={bookmark} />

                </>
            )}
            {!bookmark.duration && (
                <>
                    <td style={{ border: '1px solid black' }}>
                        <img src={bookmark.view} alt='' style={{ width: '250px', height: '150px' }} />
                    </td>
                    <td style={{ border: '1px solid black' }}>
                        <table>
                            <tr><td align='left'><li><b>URL:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.url}</td></tr>
                            <tr><td align='left'><li><b>Titre de la photo:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.title}</td></tr>
                            <tr><td align='left'><li><b>Auteur:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.author}</td></tr>
                            <tr><td align='left'><li><b>Date d'ajout dans l'application:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>
                                il y a {getTimefromDate(bookmark.addDate)}</td></tr>
                            <tr><td align='left'><li><b>Date de publication sur Flickr:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>
                            </td>
                            </tr>
                            <tr><td align='left'><li><b>Largeur x Hauteur:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.width + 'x' + bookmark.height}</td></tr>
                        </table>
                    </td>
                    <DeleteButton bookmarks={bookmarks} setBookmarks={setBookmarks} bookmark={bookmark} />
                </>
            )}
        </tr>);
}

export default BookmarkItem;