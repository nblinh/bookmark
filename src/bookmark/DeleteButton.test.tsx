
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';
import { Bookmark } from "./model";
import BookmarkItem from "./BookmarkItem";
import DeleteButton from "./DeleteButton";

const bookmarks: Bookmark[] = [
    {
        "view": "/videos/565486457",
        "url": "https://vimeo.com/565486457",
        "title": "Sylvain LhommÃ©e @ Nation Entreprenante - Episode #5",
        "author": "BARTERLINK",
        "addDate": new Date(),
        "publicDate": new Date(),
        "duration": 1070
    },
    {
        "view": "https://live.staticflickr.com/65535/51930941655_d299489ab1_b.jpg",
        "url": "https://www.flickr.com/photos/feuilllu/51930941655/in/dateposted/",
        "title": "2022 Marche Vexin",
        "author": "Pierre Metivier",
        "addDate": new Date(),
        "publicDate": new Date(),
        "width": 1024,
        "height": 683
    }
]
const setBookmarks = jest.fn();

describe('DeleteButton', () => {
    test('should render', () => {
        const { container } = render(<DeleteButton bookmarks={bookmarks} setBookmarks={setBookmarks} bookmark={bookmarks[0]} />);
        expect(container).toMatchSnapshot();
    });

    test('should have delete button', () => {
        const { container } = render(<DeleteButton bookmarks={bookmarks} setBookmarks={setBookmarks} bookmark={bookmarks[0]} />);
        expect(container.getElementsByTagName('input').item(0)?.value).toEqual('Supprimer');
    });

    test('should call update setBookmarks when click on Delete button', () => {
        const { getByTestId } = render(<BookmarkItem bookmarks={bookmarks} setBookmarks={setBookmarks} bookmark={bookmarks[0]} />);
        fireEvent.click(getByTestId('deleteButton'));
        expect(setBookmarks).toHaveBeenCalledTimes(1);
        expect(setBookmarks).toHaveBeenCalledWith([bookmarks[1]]);
    });
});