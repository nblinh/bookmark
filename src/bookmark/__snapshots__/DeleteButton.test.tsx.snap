// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteButton should render 1`] = `
<div>
  <td
    style="border: 1px solid black;"
  >
    <input
      data-testid="deleteButton"
      type="button"
      value="Supprimer"
    />
  </td>
</div>
`;
